src/it/process-jar/pom.xml

```diff
@@ -0,0 +1,115 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>org.jenkins-ci.plugins</groupId>
+    <artifactId>plugin</artifactId>
+    <version>1.424</version>
+  </parent>
+
+  <groupId>org.jenkins-ci.tools.hpi.its</groupId>
+  <artifactId>process-jar</artifactId>
+  <version>1.0-SNAPSHOT</version>
+
+  <packaging>hpi</packaging>
+
+  <name>MyNewPlugin</name>
+
+  <dependencies>
+    <dependency>
+      <groupId>com.fasterxml.jackson.core</groupId>
+      <artifactId>jackson-databind</artifactId>
+      <version>2.6.3</version>
+      <optional>true</optional>
+    </dependency>
+  </dependencies>
+
+  <build>
+    <pluginManagement>
+      <plugins>
+        <plugin>
+          <groupId>@project.groupId@</groupId>
+          <artifactId>@project.artifactId@</artifactId>
+          <version>@project.version@</version>
+        </plugin>
+        <plugin>
+          <groupId>org.kohsuke</groupId>
+          <artifactId>access-modifier-checker</artifactId>
+          <version>1.0</version>
+        </plugin>
+      </plugins>
+    </pluginManagement>
+    <plugins>
+      <plugin>
+        <groupId>@project.groupId@</groupId>
+        <artifactId>@project.artifactId@</artifactId>
+        <version>@project.version@</version>
+        <executions>
+          <execution>
+            <phase>prepare-package</phase>
+            <goals>
+              <goal>jar</goal>
+            </goals>
+            <!-- TODO the shade plugin doesn't like shading the non-main attached artifacts yet -->
+            <!--<configuration>-->
+              <!--<jarClassifier>unshaded</jarClassifier>-->
+            <!--</configuration>-->
+          </execution>
+        </executions>
+        <configuration>
+          <jarClassifier>shaded</jarClassifier>
+        </configuration>
+      </plugin>
+      <!-- TODO the shade plugin doesn't like shading the non-main attached artifacts yet -->
```

Comment:
@KostyaSha now we get to the reason why I put in the hack-alert...

So how a hpi plugin packaging *should* work is that you declare the dependency as 

```
<dependency>
  <groupId>...</groupId>
  <artifactId>...</artifactId>
  <type>hpi</type>
  <version>...</version>
</dependency>
```

The key part here is the `<type>hpi</type>`

What instead the hpi plugin does is a super painful hack... it deploys a side artifact that is of type `jar` so that when you do 

```
<dependency>
  <groupId>...</groupId>
  <artifactId>...</artifactId>
  <version>...</version>
</dependency>
```

that (due to the default value of `type` in the maven schema) becomes

```
<dependency>
  <groupId>...</groupId>
  <artifactId>...</artifactId>
  <type>jar</type>
  <version>...</version>
</dependency>
```

Which means that the jar file is added to the classpath by surefire, etc... and all your tests work...

Now in Maven land, what we really want you to do with a `.war` file is not have any java code in there and use a multi-module build to produce the `.jar` file and then just consume that in the `.war` file... if we did that for Jenkins plugins then every Jenkins plugin would need to be three modules, one for the `.jar`, one for the `.hpi` and the last one as a parent (or we could go with my suggestion of a different plugin format that is actually classpath compatible and bundles the web stuff inside the `.jar` under e.g. `META-INF/web`)

So then to unwind the above hack, the hpi plugin parses the dependencies and adds manifest entries for any that have the corresponding pom file with `<packaging>hpi</packaging>` and leaves all the rest as being bundled in the `WEB-INF/lib` unless they are `<optional>true</optional>`

So yeah crazy layer of hack on hack on hack